File: services/business-services/menu-v1/src/main/java/com/demo/bbq/business/menu/infrastructure/mapper/MenuOptionMapper.java
package com.demo.bbq.business.menu.infrastructure.mapper;

import com.demo.bbq.business.menu.domain.model.request.MenuOptionSaveRequest;
import com.demo.bbq.business.menu.domain.model.request.MenuOptionUpdateRequest;
import com.demo.bbq.business.menu.domain.model.response.MenuOption;
import com.demo.bbq.business.menu.infrastructure.repository.database.entity.MenuOptionEntity;
import com.demo.bbq.business.menu.infrastructure.repository.restclient.dto.ProductDto;
import com.demo.bbq.business.menu.infrastructure.repository.restclient.dto.ProductSaveRequestDto;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

@Mapper(componentModel = "spring")
public interface MenuOptionMapper {

  MenuOption fromEntityToResponse(MenuOptionEntity menuOption);
  MenuOptionEntity fromSaveRequestToEntity(MenuOptionSaveRequest menuOption);

  MenuOptionEntity fromUpdateRequestToEntity(MenuOptionUpdateRequest menuOption, String productCode);

  @Mapping(target = "productCode", source = "product.code")
  MenuOption fromProductToResponse(MenuOptionEntity menuOption, ProductDto product);

  @Mapping(target = "code", source = "menuOption.productCode")
  ProductSaveRequestDto fromSaveRequestToProduct(MenuOptionSaveRequest menuOption, String scope);

  ProductSaveRequestDto fromUpdateRequestToProduct(MenuOptionUpdateRequest menuOption, String scope);
}

File: services/business-services/menu-v3/infrastructure/mapper/menu_option_mapper.go
package mapper

import (
	"github.com/mitchellh/mapstructure"
	"menu-v3/domain/model/request"
	"menu-v3/domain/model/response"
	"menu-v3/infrastructure/repository/entity"
)

func RequestToEntity(request request.MenuOptionRequest) entity.MenuOption {
	var menuOption entity.MenuOption
	mapstructure.Decode(request, &menuOption)
	return menuOption
}

func UpdateRequestToEntity(request request.MenuOptionRequest, menuOption *entity.MenuOption) {
	mapstructure.Decode(request, menuOption)
}

func EntityToResponse(menuOption entity.MenuOption) response.MenuOptionResponse {
	var response response.MenuOptionResponse
	mapstructure.Decode(menuOption, &response)
	return response
}
